android/guava/src/com/google/common/cache/LocalCache.java

```diff
@@ -2201,7 +2196,20 @@ V waitForLoadingValue(ReferenceEntry<K, V> e, K key, ValueReference<K, V> valueR
         throw new AssertionError();
       }
 
-      checkState(!Thread.holdsLock(e), "Recursive load of: %s", key);
+      // As of this writing, the only prod ValueReference implementation for which isLoading() is
+      // true is LoadingValueReference. However, that might change, and we already have a *test*
+      // implementation for which it doesn't hold. So we check instanceof to be safe.
```

Comment:
Whoa, this is complicated! Thanks.

Hmm, does the right thing even currently happen if someone starts a load when `compute` is already running? If `ComputingValueReference` returns `false` from `isLoading`, then would we [call `valueReference.get()`](https://github.com/google/guava/blob/f4c1264c9f8ee1e72af39c4fa3776b86b8519ed1/guava/src/com/google/common/cache/LocalCache.java#L2144C15-L2144C45)... and then either get back `null` (if there was no previous value) and say an entry was `COLLECTED` (which seems iffy) or get back the previous value (which might be fine but which I haven't thought about)? We probably never get to `waitForLoadingValue` and the deadlock detection at all? (If we did, we'd trigger [the `AssertionError` above](https://github.com/google/guava/blob/f4c1264c9f8ee1e72af39c4fa3776b86b8519ed1/guava/src/com/google/common/cache/LocalCache.java#L2204-L2206)!)

Maybe there's sketchy stuff going on but this PR at least preserves existing behavior?

Reply:
https://github.com/google/guava/issues/5439? https://github.com/google/guava/issues/3719? I am going to try not to get sucked much deeper into this. Use Caffeine, everyone! (And hopefully we can still make this "simple" virtual-thread fix :))

Reply:
yeah, there are lots of little gotchas and I may have introduced some in my fix attempts too. I'd only suggest modifying that comment and certainly not expand the scope. Even Doug had challenges with recursive computes and there are so many interactions that I couldn't possibly keep them all in my head (hence 10M test executions per build in hopes to catch regressions).

Reply:
Sounds good. I have tweaked the comment to avoid implying that `instanceof LoadingValueReference` implies `isLoading()`.

Reply:
> Hmm, does the right thing even currently happen if someone starts a load when compute is already running?

oh, to answer your question this cannot happen because the `compute` methods are implemented very coarsely. They hold the segment lock during the computation which blocks all other writes to that segment, meaning a loader would wait to obtain the lock and by then the computing value reference would be gone.

The alternative would be to use retry loops on all write paths by taking the segment lock, if loading then waiting on the entry to complete outside of the segment lock, or else perform its entry action accordingly. That would have been far more invasive and may not have maintained linearizability, so I believe simpler approach was used just to avoid accidents as the default method version is non-atomic.

ConcurrentHashMap uses fine-grained segments that are dynamically added, so it similarly locks rather than using retry loops. Caffeine delegates loading to computeIfAbsent as merely an alias, where the only negative effect is long-running computes could impact siblings. Thus, AsyncCache mitigates that when an issue since it sacrifices linearization of the loading value (only of the future mapping) so that the map operations are fast, and the synchronous view uses the retry loops approach as best-effort (not linearizable as it fails Lincheck).

Reply:
Oh, good. I'm glad that we at least do not have an additional bug on top of the one you already reported... :)

Reply:
Hey, can you give me an update regarding the current status of this PR?

Reply:
Hi, sorry.

The testing results are good. My reviewer for the internal change had some other things come up, so I'll reassign the review to someone else.

While we're here: How important is it to you for us to publish a release that contains this change? I'm hoping to publish one soonish no matter what, but it would be useful to know specifically about this change.